### PR TITLE
Pump It Up

### PR DESCRIPTION
**Pull request by mjmagee991**
Added simple heat pump driver
Created sample to demonstrate behavior and allow easy testing

### CODE REVIEW

File: src/TMS/dev/HeatPump.cpp [hunk 1]
```diff
@@ -0,0 +1,36 @@
+#include <TMS/dev/HeatPump.hpp>
+
+namespace TMS {
+
+HeatPump::HeatPump(IO::PWM& pwm, uint8_t initSpeed) : pwm(pwm) {
```
Comment (cbolles) [inline src/TMS/dev/HeatPump.cpp]: For the `initSpeed` when the TMS system starts up, we do not want the pump running, so the pump should start out stopped.

**Review by cbolles — CHANGES_REQUESTED** [review]
Just need to make sure the pump can start up without running

File: src/TMS/dev/HeatPump.cpp [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    // Ensures initSpeed falls in the required range
+    if (initSpeed < MIN_INIT_SPEED) {
+        this->initSpeed = MIN_INIT_SPEED;
+    } else if (initSpeed > MAX_SPEED) {
+        this->initSpeed = MAX_SPEED;
+    } else {
+        this->initSpeed = initSpeed;
+    }
```
Comment (ddp5730) [inline src/TMS/dev/HeatPump.cpp]: MIN_INIT_SPEED is 7.  But initSpeed is supposed to be ranged from 0-100.  Shouldn't this check if initSpeed < 0 here.

File: src/TMS/dev/HeatPump.cpp [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+void HeatPump::setSpeed(uint8_t speed) {
+    // Ensures speed falls in the required range
+    if (speed > MAX_SPEED) {
+        speed = MAX_SPEED;
+    }
+
+    float dutyCycle = SPEED_TO_DUTY_CYCLE(speed);
+    pwm.setDutyCycle(dutyCycle);
+}
```
Comment (ddp5730) [inline src/TMS/dev/HeatPump.cpp]: Do you apply the 3 ms pulse here when setting a speed after it was previously stopped?

**Review by ddp5730 — CHANGES_REQUESTED** [review]

File: src/TMS/dev/HeatPump.cpp (hunk 1 above)
Comment (mjmagee991) [reply src/TMS/dev/HeatPump.cpp]: It's now been fully removed.

File: src/TMS/dev/HeatPump.cpp (hunk 2 above)
Comment (mjmagee991) [reply src/TMS/dev/HeatPump.cpp]: The MIN_INIT_SPEED had to be 7 to make sure the duty cycle and period result in an impulse of >= 3 ms. I changed some things around and added documentation to make this clearer. Let me know if anything is still vague.

File: src/TMS/dev/HeatPump.cpp (hunk 3 above)
Comment (mjmagee991) [reply src/TMS/dev/HeatPump.cpp]: This is related to the minimum initialization speed. In the new implementation, the impulse is taken care of in this method. Let me know if that is clear now.

**Review by cbolles — APPROVED** [review]
Looks good!

**Review by ddp5730 — APPROVED** [review]
It makes more sense now and everything looks clean.  Nice work!

**Pull request merged**